lib/srv/db/mysql/proxy.go

```diff
@@ -81,7 +85,22 @@ func (p *Proxy) HandleConnection(ctx context.Context, clientConn net.Conn) (err
 	if err != nil {
 		return trace.Wrap(err)
 	}
-	serviceConn, authContext, err := p.Service.Connect(ctx, server.GetUser(), server.GetDatabase())
+	clientIP, err := utils.ClientIPFromConn(clientConn)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Apply rate limiting.
+	if err := p.Limiter.AcquireConnection(clientIP); err != nil {
+		return trace.LimitExceeded("client %v exceeded connection limit", clientIP)
+	}
+	defer p.Limiter.ReleaseConnection(clientIP)
+
+	serviceConn, authContext, err := p.Service.Connect(ctx, common.ConnectParams{
+		User:     server.GetUser(),
+		Database: server.GetDatabase(),
+		ClientIP: clientIP,
+	})
```

Comment:
Cane we move this logic at the top of the `HandleConnection` function to make sure that the limiter applies before proxy logic is invoked ?

Reply:
Yes, I agree. We should cut the connection as early as possible. The only downside of that is that right now error returned to a client says something random as EOF or incorrect startup sequence instead of "exceeded connection limit". I'm not sure how big problem is that tho.

Reply:
> The only downside of that is that right now error returned to a client says something random as EOF or incorrect startup sequence instead of "exceeded connection limit"

@jakule I'd say this is definitely not a great UX. We should strive to report all errors clearly and accurately. For MySQL proxy specifically, there is code that handles returning errors to clients in proper format:

https://github.com/gravitational/teleport/blob/v8.0.6/lib/srv/db/mysql/proxy.go#L63-L73

Would the client get a proper error in case of rate limiting if we do this defer before invoking the limiter?

Reply:
@r0mant As we talked I updated the logic to return a more human friendly error message. Here is an example:
```
psql: error: connection to server at "localhost" (::1), port 58128 failed: Connection refused
	Is the server running on that host and accepting TCP/IP connections?
connection to server at "localhost" (127.0.0.1), port 58128 failed: exceeded connection limit
ERROR: exit status 2
```
@smallinsky The logic has been moved to a different place to return more descriptive error than before. I made the changes in the proxy and the server.